datalad/core/dataset/gitrepo.py

```diff
@@ -0,0 +1,531 @@
+
+        Parameters
+        ----------
+        args : list of str
+          Arguments to pass to `git`.
+        files : list of str, optional
+          File arguments to pass to `git`. The advantage of passing these here
+          rather than as part of `args` is that the call will be split into
+          multiple calls to avoid exceeding the maximum command line length.
+        expect_stderr : bool, optional
+          Standard error is expected and should not be elevated above the DEBUG
+          level.
+        sep : str, optional
+          Split the output by `str.split(sep)` rather than `str.splitlines`.
+
+        Raises
+        ------
+        CommandError if the call exits with a non-zero status.
+        AssertionError if there is more than one line of output.
+        """
+        lines = list(self.call_git_items_(args, files=files,
+                                          expect_stderr=expect_stderr))
+        if len(lines) > 1:
+            raise AssertionError(
+                "Expected {} to return single line, but it returned {}"
+                .format(["git"] + args, lines))
+        return lines[0]
+
+    def call_git_success(self, args, files=None, expect_stderr=False):
+        """Call git and return true if the call exit code of 0.
+
+        Parameters
+        ----------
+        args : list of str
+          Arguments to pass to `git`.
+        files : list of str, optional
+          File arguments to pass to `git`. The advantage of passing these here
+          rather than as part of `args` is that the call will be split into
+          multiple calls to avoid exceeding the maximum command line length.
+        expect_stderr : bool, optional
+          Standard error is expected and should not be elevated above the DEBUG
+          level.
+
+        Returns
+        -------
+        bool
+        """
+        try:
+            self._call_git(
+                args, files, expect_fail=True, expect_stderr=expect_stderr)
+        except CommandError:
+            return False
+        return True
+
+    def add_fake_dates(self, env):
+        """Add fake dates to `env`.
+
+        Parameters
+        ----------
+        env : dict or None
+            Environment variables.
+
+        Returns
+        -------
+        A dict (copied from env), with date-related environment
+        variables for git and git-annex set.
+        """
+        env = (env if env is not None else environ).copy()
+        if 'DATALAD_FAKE_DATE' in env:
+            # there is already a fake date setup
+            return env
+        # prevent infinite recursion, should this function be called again
+        # inside the next call.
+        environ['DATALAD_FAKE_DATE'] = '1'
+        last_date = list(self.for_each_ref_(
+            fields='committerdate:raw',
+            count=1,
+            pattern='refs/heads',
+            sort="-committerdate",
+        ))
+        del environ['DATALAD_FAKE_DATE']
```

Comment:
In case a Python caller happens to catch an exception (most likely unintentionally) that is raised by `for_each_ref_`, it might be worth guarding this "DATALAD_FAKE_DATE" set/delete with a try-finally.

Reply:
Done